### DISCUSSION TITLE
TCP Server Led Toggle from HMI

### DISCUSSION THREAD

**Discussion by rvxfahim** [eModbus/eModbus#216] [Q&A] [answered]
I am trying to figure out how to do a simple program like toggling an LED of ESP32 from HMI.
My idea is that ESP32 will have a register which will hold a integer value. On boot ESP32 will write 0 to a register, and later if HMI writes 1 to this register then ESP32 will do digitalWrite(LedPin).

Things that I am trying to find out:
How to add a register in ESP32
how to write to this register from inside loop
how to read a register inside the HMI.

**Comment by Miq1** [comment, marked as answer]
If I understand it correctly, the ESP32 is to drive the LED and is connected over Modbus TCP to have the LED lit or off? Then it is regardless who is requesting the ESP32 (acting as a Modbus TCP server) - can be an HMI or something completely different.

You will want to create a ModbusServerTCP instance in the ESP32 sketch and then register two worker functions for function codes 0x03 READ_HOLD_REGISTER and 0x06 WRITE_HOLD_REGISTER. Let us assume your chosen register address will be 7, the ESP32 does not respond to any other address and the LED is connected to some LED_GPIO you have chosen.

Then the function serving the 0x03 function code may look similar to
```
ModbusMessage FC03(ModbusMessage request) {
  uint16_t address;           // requested register address
  ModbusMessage response;     // response message to be sent back

  // get request values
  request.get(2, address);
  request.get(4, words);

  // Address equals 7 and words is 1?
  if (address == 7 && words == 1) {
    // Looks okay. Set up message with serverID, FC, length of data and the LED state
    response.add(request.getServerID(), request.getFunctionCode(), 2, (uint16_t)(digitalRead(LED_GPIO) ? 1 : 0));
  } else {
    // No, either address or words are outside the limits. Set up error response.
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  }
  return response;
}
```
Note that there is no register at all, we need no separate value for your task to store.

The 0x06 function is similar:
```
ModbusMessage FC06(ModbusMessage request) {
  uint16_t address;           // requested register address
  uint16_t data;             // value to be set
  ModbusMessage response;     // response message to be sent back

  // get request values
  request.get(2, address);
  request.get(4, data);

  // Address equals 7?
  if (address == 7) {
    // Depending on data being 0 or not, toggle LED
    digitalWrite(LED_GPIO, (data ? HIGH : LOW));
    // Return echo response
    return ECHO_RESPONSE;
  } 
  // No, address is outside the limits. Set up error response.
  response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  return response;
}
```
Now anything written to register address 7 that is >0 will light the LED, a value of 0 will darken it.

**Reply by rvxfahim** [reply]
Thanks a lot, I was stumbling around the whole concept of having no registers at all, because coming from other libraries where they add registers to server and then those registers are written by a client and then declaring some callback functions for register value-change.

**Reply by Miq1** [reply]
The Modbus "register" is a kind of virtual concept to enumerate data for the outside world only. If you choose to really implement it by a section of memory holding values or f.i. just use it to trigger functions for each number returning some value or do it completely different again is all up to you.
